modules/layers/src/text-layer/font-atlas-manager.ts

```diff
@@ -188,26 +197,98 @@ export default class FontAtlasManager {
     }
 
     // update fontAtlas with new settings
-    const fontAtlas = this._generateFontAtlas(this._key, charSet, cachedFontAtlas);
+    const fontAtlas = this._generateFontAtlas(charSet, cachedFontAtlas);
     this._atlas = fontAtlas;
 
     // update cache
     cache.set(this._key, fontAtlas);
   }
 
-  private _generateFontAtlas(
-    key: string,
-    characterSet: Set<string>,
-    cachedFontAtlas?: FontAtlas
-  ): FontAtlas {
-    const {fontFamily, fontWeight, fontSize, buffer, sdf, radius, cutoff} = this.props;
+  private _generateFontAtlas(characterSet: Set<string>, cachedFontAtlas?: FontAtlas): FontAtlas {
     let canvas = cachedFontAtlas && cachedFontAtlas.data;
     if (!canvas) {
       canvas = document.createElement('canvas');
       canvas.width = MAX_CANVAS_WIDTH;
     }
+
     const ctx = canvas.getContext('2d')!;
 
+    return this.props.sdf
+      ? this._generateSDFFontAtlas(canvas, ctx, characterSet)
+      : this._generateRegularFontAtlas(canvas, ctx, characterSet, cachedFontAtlas);
+  }
+
+  private _generateSDFFontAtlas(
+    canvas: HTMLCanvasElement,
+    ctx: CanvasRenderingContext2D,
+    characterSet: Set<string>
+  ): FontAtlas {
+    const {fontFamily, fontWeight, fontSize, buffer, radius, cutoff} = this.props;
+    const mapping: CharacterMapping = {};
+    const tinySDF = new TinySDF({
+      fontSize,
+      buffer,
+      radius,
+      cutoff,
+      fontFamily,
+      fontWeight: `${fontWeight}`
+    });
+    const cellSize = fontSize + buffer * 2;
+    const layoutRowHeight = fontSize * HEIGHT_SCALE;
+    const columns = Math.floor(canvas.width / cellSize);
+    const rows = Math.ceil(characterSet.size / columns);
+    const requiredCanvasHeight = nextPowOfTwo(rows * cellSize);
+
+    if (canvas.height !== requiredCanvasHeight) {
+      resizeCanvas(ctx, canvas.width, requiredCanvasHeight);
+    }
+
+    let x = 0;
+    let y = 0;
+    for (const char of characterSet) {
+      const {data, width, height, glyphTop, glyphAdvance} = tinySDF.draw(char);
+
+      const imageData = ctx.createImageData(width, height);
+      populateAlphaChannel(data, imageData);
+      ctx.putImageData(imageData, x, y);
+
+      mapping[char] = {
+        x,
+        y,
+        // use glyphAdvance over glyphWidth due to whitespace characters
+        width: glyphAdvance,
+        textureHeight: cellSize,
+        textureWidth: width,
+        textureOffsetY: fontSize - glyphTop
```

Comment:
I find the names `texture*` confusing (since everything is in a texture). Going back to my diagram, `x` `y` `width` and `height` should describe the green box, and we should only need two more fields `layoutWidth` and `layoutHeight` to describe the red box.

Reply:
I've removed the `texture*` prefix so there is now; `x`, `y`, `width`, `height`, `layoutWidth`, & `layoutOffsetY`.

The reason I've used `layoutOffsetY` is because the vertical alignment of the font is no longer baked into the SDF atlas as with the regular atlas, instead the characters are rendered from the very top of their allocated "cell" and the vertical alignment is fixed when positioning the icons instead. The reason I did this was because aligning them within their cell would often cause dangling letters like "y" to have their gradient clipped.

How the atlas looks without the alignment baked in. The green outline being the "render bounds" & the red being the "layout" bounds.
![image](https://user-images.githubusercontent.com/1169866/187954700-b2cfd822-4e48-45e3-8923-a19a8732433b.png)

Here's how it looks with the alignment being baked in, notice the "y" and a few other get their gradient clipped.
![image](https://user-images.githubusercontent.com/1169866/187956000-8b918ccf-45c4-4774-8a85-d117e01c03c6.png)

Not baking in the alignment would also allow for each character to have a far smaller texture height due to not requiring the whitespace, however when I try this for some reason it completely breaks the positioning of the character icons, I've not figured out why. Also having varying height character icons would also require them to have their scaled individually.